### PR TITLE
feat: support HttpClient configuration

### PR DESCRIPTION
**Pull request by greathouse**
## Description

- Allow customizing the HttpClient instance. There is an open bug here. https://github.com/OpenAPITools/openapi-generator/issues/11943. I tried the `genericHost` option as described in the conversation, but that had problems too (and is still marked as "experimental").
- Allow clients to modify request options (ie: adding `BT-TRACE-ID` and `BT-IDEMPOTENCY-KEY`). The original generated template did not pass in a `RequestOptions` class as parameters. I applied the patch found here: https://github.com/OpenAPITools/openapi-generator/issues/11010. I had to make some minor modifications.
- Add `CorrelationId` and `IdempotencyKey` to the `RequestOptions` as shortcuts to adding the request headers.
- Ensure that RestSharp throws on every error. This is to support custom HttpHandlers that throw exceptions like the `TenantUserAuthenticationHandler` in the Admin API.

Draft PR for the Developer Documentation: https://github.com/Basis-Theory/developers.basistheory.com/pull/409

## Testing required outside of automated testing?

- [ ] Not Applicable

### Screenshots (if appropriate):

- [ ] Not Applicable

## Rollback / Rollforward Procedure

- [ ] Roll Forward
- [ ] Roll Back

## Reviewer Checklist

- [ ] Description of Change
- [ ] Description of outside testing if applicable.
- [ ] Description of Roll Forward / Backward Procedure
- [ ] Documentation updated for Change

### CODE REVIEW

File: .openapi-generator/templates/ApiClient.mustache [hunk 1]
```diff
@@ -172,6 +173,7 @@ namespace {{packageName}}.Client
     {{>visibility}} partial class ApiClient : ISynchronousClient{{#supportsAsync}}, IAsynchronousClient{{/supportsAsync}}
     {
         private readonly string _baseUrl;
+        private HttpClient _httpClient;
```
Comment (greathouse) [inline .openapi-generator/templates/ApiClient.mustache:176]: Allowing the client to customize the HttpClient that RestSharp uses internally.

File: .openapi-generator/templates/ApiClient.mustache [hunk 2]
```diff
@@ -451,13 +470,13 @@ namespace {{packageName}}.Client
             var clientOptions = new RestClientOptions(baseUrl)
             {
                 ClientCertificates = configuration.ClientCertificates,
-                CookieContainer = cookies,
                 MaxTimeout = configuration.Timeout,
                 Proxy = configuration.Proxy,
-                UserAgent = configuration.UserAgent
+                UserAgent = configuration.UserAgent,
+                ThrowOnAnyError = true
```
Comment (greathouse) [inline .openapi-generator/templates/ApiClient.mustache:476]: Throws all errors encountered during the http process.

File: .openapi-generator/templates/RequestOptions.mustache [hunk 3]
```diff
@@ -60,6 +60,24 @@ namespace {{packageName}}.Client
...
+    
+        public String IdempotencyKey
+        {
+            set
+            {
+                this.HeaderParameters.Remove("BT-IDEMPOTENCY-KEY");
+                this.HeaderParameters.Add("BT-IDEMPOTENCY-KEY", value);
+            }
+        }
+
```
Comment (greathouse) [inline .openapi-generator/templates/RequestOptions.mustache:80]: Convenience properties to set Basis Theory specific request headers.

File: src/BasisTheory.net/Api/PermissionsApi.cs [hunk 4]
```diff
@@ -219,10 +223,16 @@ public List<Permission> Get(string applicationType = default(string), int operat
...
+        /// <param name="requestOptions"></param>
         /// <returns>ApiResponse of List&lt;Permission&gt;</returns>
-        public BasisTheory.net.Client.ApiResponse<List<Permission>> GetWithHttpInfo(string applicationType = default(string), int operationIndex = 0)
+        public BasisTheory.net.Client.ApiResponse<List<Permission>> GetWithHttpInfo(string applicationType = default(string), int operationIndex = 0, BasisTheory.net.Client.RequestOptions requestOptions = default(BasisTheory.net.Client.RequestOptions))
         {
-            BasisTheory.net.Client.RequestOptions localVarRequestOptions = new BasisTheory.net.Client.RequestOptions();
+            if(requestOptions == default(BasisTheory.net.Client.RequestOptions))
+            {
+                requestOptions = new BasisTheory.net.Client.RequestOptions();
+            }
```
Comment (greathouse) [inline src/BasisTheory.net/Api/PermissionsApi.cs:233]: Allowing the client to pass a `RequestOptions` object. The previous template created the RequestOptions internally.

**Review by dhudec — APPROVED** [review]

**Pull request merged**